### ISSUE TITLE
list.append seems to break return type checker

### ISSUE THREAD

**Issue by AugPro** [python/mypy#8734]
Hello, concerning the function below:
```
def foo() -> List[int]:
    listings = None
    if 2 != 2:
        listings = list()
        listings.append(1)
    return listings
```

I expected: 
`Incompatible return value type (got "Optional[List[int]]", expected "List[int]") `

I got no error

This one does throw the expected error (same config):
```
def foo() -> List[int]:
    listings = None
    if 2 != 2:
        listings = [1]
    return listings
```

running mypy `0.770` on python `3.8`

I would appreciate it if you could tell me if I am doing something wrong :)

Have a good day !

**Comment by msullivan** [comment]
Yeah that looks like a bug. Something must be wrong with the case where we try to refine a partial type multiple times in a conditional way?

**Comment by AugPro** [comment]
I've never contributed to this repo but if you want I'd be glad to help!
